src/main/java/io/jenkins/plugins/httpclient/RobustHTTPClient.java

```diff
@@ -0,0 +1,281 @@
+import jenkins.MasterToSlaveFileCallable;
+import jenkins.security.MasterToSlaveCallable;
+import jenkins.util.JenkinsJVM;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.StatusLine;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpPut;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.entity.FileEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.Beta;
+
+/**
+ * Utility to make HTTP connections with protection against transient failures.
+ */
+@Restricted(Beta.class)
+public final class RobustHTTPClient implements Serializable {
+
+    private static final long serialVersionUID = 1;
+
+    private static final ExecutorService executors = JenkinsJVM.isJenkinsJVM() ? Computer.threadPoolForRemoting : Executors.newCachedThreadPool();
+
+    private int stopAfterAttemptNumber;
+    private long waitMultiplier;
+    private long waitMaximum;
+    private long timeout;
+
+    /**
+     * Creates a client configured with reasonable defaults from system properties.
+     * <p>This constructor should be run in the Jenkins master.
+     * To make requests from an agent JVM, create a {@code final} field of this type in your {@link MasterToSlaveCallable} or similar;
+     * set it with a field initializer (run in the callable’s constructor on the master),
+     * letting the agent deserialize the configuration.
+     */
+    public RobustHTTPClient() {
+        JenkinsJVM.checkJenkinsJVM();
+        this.stopAfterAttemptNumber = Integer.getInteger(RobustHTTPClient.class.getName() + ".STOP_AFTER_ATTEMPT_NUMBER", 10);
+        this.waitMultiplier = Long.getLong(RobustHTTPClient.class.getName() + ".WAIT_MULTIPLIER", 100);
+        this.waitMaximum = Long.getLong(RobustHTTPClient.class.getName() + ".WAIT_MAXIMUM", 300);
+        this.timeout = Long.getLong(RobustHTTPClient.class.getName() + ".TIMEOUT", /* 15m */15 * 60);
+    }
+
+    /**
+     * Number of upload/download attempts of nonfatal errors before giving up.
+     */
+    public void setStopAfterAttemptNumber(int stopAfterAttemptNumber) {
+        this.stopAfterAttemptNumber = stopAfterAttemptNumber;
+    }
+
+    /**
+     * Initial number of milliseconds between first and second upload/download attempts.
+     * Subsequent ones increase exponentially.
+     * Note that this is not a <em>randomized</em> exponential backoff;
+     * and the base of the exponent is currently hard-coded to 2.
+     */
+    public void setWaitMultiplier(long waitMultiplier) {
+        this.waitMultiplier = waitMultiplier;
+    }
+
+    /**
+     * Maximum number of seconds between upload/download attempts.
```

Comment:
nit: I think it would be good to use the same units or make the user pass a `ChronoUnit`/`TimeUnit` for `waitMultiplier` and `waitMaximum` (maybe also `timeout`) to avoid potential confusion.